### PR TITLE
Added the C/C++ implementation for the root-finder algorithm using the tangent method

### PR DESCRIPTION
**Pull request by HeshamElsaman**
# The C/C++ implementation for root-finding

## Linting
- Supposedly, it 100.0% lints.

## Content
- Added the function to find the root utilizing the c-types.
- Files Changed: calculus.py, test_calculus.py
- Files Added: tangent_root_finder.cpp, tangent_root_finder.dll

## Reviwer:
- @haiderabbas007

### CODE REVIEW

**Comment by HeshamElsaman** [comment]
@haiderabbas007 Apologies for the late notice, but I hope you can review it as soon as possible. Appreciated!

**Comment by HeshamElsaman** [comment]
@zbpetersbuf Should I be adding you as a Reviewer to this? or should I ask in #37 if someone still needs a review to get the points for it?

**Comment by Cosmos491** [comment]
> @zbpetersbuf Should I be adding you as a Reviewer to this? or should I ask in #37 if someone still needs a review to get the points for it?

I can review it, if you want. @HeshamElsaman

**Comment by HeshamElsaman** [comment]
> I can review it, if you want. @HeshamElsaman

@Cosmos491 Sure, NP!

**Comment by Cosmos491** [comment]
Sure, I am on it

**Comment by haiderabbas007** [comment]
I need to review two PRs, and have done only 1 so far. @Cosmos491 you can go ahead and review it. I have an exam this morning and will get free later. I will provide my general feedback to serve as a "review" when I get free, but of course the merging can be done once cosmos491 approves.

**Comment by HeshamElsaman** [comment]
@haiderabbas007 good luck with your exam, wish you the best!

**Comment by Cosmos491** [comment]
You passed all the checks, however I do have some suggestions:
1. For the cpp file, in case the method does not converge within the maximum number of iterations, you might benefit from logging or additional information in the Result structure, such as the final value of x and the number of iterations completed.
2. Parameters like f, fprime, and x0 could be marked as const to convey that they are not modified.
3. In test_calculus, you can add assertions and mock functions.

Let me know what you think @HeshamElsaman

**Comment by HeshamElsaman** [comment]
> 1. For the cpp file, in case the method does not converge within the maximum number of iterations, you might benefit from logging or additional information in the Result structure, such as the final value of x and the number of iterations completed.
> 2. Parameters like f, fprime, and x0 could be marked as const to convey that they are not modified.
> 3. In test_calculus, you can add assertions and mock functions.

Thanks for your suggestions.

1. I guess I can add that but I guess it would be just for the sake of enhancing the functionality.
2. They're not supposed to be constants. Could you, please, elaborate more on what you mean?
3. I can do so, but the same goes here as in 1 for it would only be for the sake of enhancing the code.

If there's nothing else than that, I guess I can work on your suggestions later today. It would be better though to merge this first if there's no critical issues with it as long as it adheres to the requirements of this homework if you believe so.

**Comment by Cosmos491** [comment]
> > 1. For the cpp file, in case the method does not converge within the maximum number of iterations, you might benefit from logging or additional information in the Result structure, such as the final value of x and the number of iterations completed.
> > 2. Parameters like f, fprime, and x0 could be marked as const to convey that they are not modified.
> > 3. In test_calculus, you can add assertions and mock functions.
> 
> Thanks for your suggestions!
> 
> 1. I guess I can add that but I guess it would be just for the sake of enhancing the functionality.
> 2. They're not supposed to be constants. Could you, please, elaborate more on what you mean?
> 3. I can do so, but the same goes here as in 1 for it would only be for the sake of enhancing the code.
> 
> If there's nothing else than that, I guess I can work on your suggestions later today. It would be better though to merge this first if there are no critical issues with it as long as it adheres to the requirements of this homework if you believe so.

Oh, yeah they are not constant, my bad. Yeah my suggestions were more towards enhancement. I can approve the PR now

**Review by Cosmos491 — APPROVED** [review]

**Comment by HeshamElsaman** [comment]
@Cosmos491 Thanks a lot for your input! I'll start working on your suggestions later today once I find some available time.

**Comment by zbpetersbuf** [comment]
just @ me when you are complete with your work

**Comment by zbpetersbuf** [comment]
@laserlab how to fix the too many lines on pylint? its in workflows right?

pylint --max-module-lines=1500 calculus.py

**Comment by zbpetersbuf** [comment]
@haiderabbas007 please approve the review, if you do not do this by either Tuesday maybe Wednesday if @HeshamElsaman allows it this pull will be merged, if you do not approve it before then you will not get the point for it

**Comment by HeshamElsaman** [comment]
@Cosmos491 I added the information you suggested in your first suggestion. Unfortunately, due to time constraints I won't be able to add your third suggestion, apologies for that. But I've tested for the most important issues I believe, so I hope this will be fine.

**Comment by HeshamElsaman** [comment]
@zbpetersbuf could we remove the dummy function and its test?

**Comment by HeshamElsaman** [comment]
@haiderabbas007 could you, please, review this as soon as possible? Appreciated!

**Comment by zbpetersbuf** [comment]
@HeshamElsaman I guess you can delete it if you want to, is it cause of the length? I can force merge it regardless of this, also do you want me to merge your pull request before tomorrow? I feel like this isnt on you for having a late hw cause @haiderabbas007 isn't finalizing their review, and since you already have one approval you dont need theirs

**Review by haiderabbas007 — APPROVED** [review]

**Comment by haiderabbas007** [comment]
The code is functional, and there are no issues with operationality. I have approved this PR. I do have a few suggestions though, which will only make it more of a masterpiece which it already is:

- For the C++ implementation, the function stops immediately when `f'(x)` is near zero but doesn’t provide detailed feedback. You might want to add a field in the `Result` struct to indicate whether the failure was due to division by zero or reaching the maximum iterations.
- Default parameters in the `cpp_root_tangent function` (e.g., `double tol = 1e-6`) are unnecessary since these values are likely set in Python. Removing the default values from the C++ signature could avoid confusion.
- For your Python code `cpp_root_tangent`, while docstrings are present, they lack important details such as example usage or a description of outputs, so you might want to enhance them.
- The DLL path in `root_cpp = ctypes.CDLL("./tangent_root_finder.dll")` is hardcoded, which can cause portability issues. Using `os.path` or `pathlib` would construct a platform-independent path.
- The tests use `pytest.raises` but the error messages in exceptions could be clearer. Include more context in raised exceptions, e.g., the specific reason for convergence failure.
- Add tests for edge cases like a flat function $f(x)=c$, where no root exists and a derivative close to zero but non-zero, to verify stability.

Overall, well done!

File: calculus.py
Comment (zbpetersbuf) [inline calculus.py:79]: # pylint: disable=C0302 to remove the too many lines error

**Review by zbpetersbuf — APPROVED** [review]

**Pull request merged**
